pup-0003.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Cherry-Picking Commits
+author: David Davis
+created: 01-Apr-2017
+status:
+---
+
+## Summary
+
+Currently in Pulp, we merge features to master. However, hotfixes and bug fixes
+are merged to release branches which then are merged to newer release branches
+and then to master (a process called "merging forward"). More info about
+merging forward can be found in our documentation:
+
+http://docs.pulpproject.org/en/2.12/dev-guide/contributing/merging.html#merging-your-changes
+
+Also, a good illustration of this process is outlined in the following webpage:
+
+http://nvie.com/posts/a-successful-git-branching-model/
+
+This PUP is to use cherry-picking instead of merging forward. Using this model,
+PRs would be made against master and then changes would be cherry-picked to
+release branches as needed. For a more detailed workflow, see the Design
+section.
+
+
+## Motivation
+
+There are a few motivating factors for this proposal. First, our git history is
+complex with multiple branches feeding into one another and numerous merge
+commits. Using cherry-picks would simplify our git history by keeping it more
+linear and eliminating merge commits.
+
+Secondly, opening a PR requires knowledge of our current process. This can
+create a barrier to entry for community members. With the cherry-pick model,
+all PRs are opened against master and the onus to update release branches is on
+the person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+Another minor caveat is that changes that we want to only be made to older
+release branches (but not to newer ones) still need to be merged forward. We
+could elminate the need for this with cherry-picking. It's worth noting that
+this rarely happens.
+
+Lastly, we've also seen some mistakes where branches are accidentally merged
+into other branches. Commits could be accidentally cherry-picked but this would
+be less disastrous since it's only a single commit.
+
+
+## Detailed Design
+
+This proposal would keep the current branching scheme [1] involving master,
+x.y-dev branches, and x.y-release branhces. The workflow by which changes are
+moved between the branches would however change.
+
+### Story
+
+1. All story PRs are opened and merged against master.
+
+### Bug Fix
+
+1. All bug fix PRs are also opened against master
+2. After the PR is ready to be merged, the person merging the commit would
+   first merge the commit to master. Then, this person would cherry-pick the
+   change from master to the latest x.y-dev branch.
```

Comment:
Bugfixes *should* go to the current bugfix release stream, but if it needs conflict resolution or reworking of the implementation then it's probably not worth it unless the bug is very important. What qualifies as very important is subjective, but as an absolute policy I don't think we need 100% of bugs to land on the bugfix stream. I think that time could be better spent fixing more bugs versus fixing the same bug twice so users can receive the fix a few weeks earlier.